File: dumper.lua
-- CS:S v93 x64 External Netvar Dumper (CE Lua)
-- ===============================================
-- Paste into CE Lua Engine (Ctrl+L), Execute. No Python, no relay.
-- Writes netvar_dump.txt to your Desktop.
--
-- PREREQUISITES: CE attached to cstrike_win64.exe, game in an active map.

-- ---- struct offsets (Source SDK x64) ----
local CC_NETWORKNAME = 0x10  -- const char* m_pNetworkName
local CC_RECVTABLE    = 0x18  -- RecvTable* m_pRecvTable
local CC_NEXT         = 0x20  -- ClientClass* m_pNext
local CC_CLASSID      = 0x28  -- int m_ClassID

local RT_PROPS        = 0x00  -- RecvProp* m_pProps
local RT_NPROPS       = 0x08  -- int m_nProps
local RT_NETTABLENAME = 0x18  -- const char* m_pNetTableName

local RP_VARNAME      = 0x00  -- const char* m_pVarName
local RP_RECVTYPE     = 0x08  -- int m_RecvType
local RP_OFFSET       = 0x4C  -- int m_Offset (ESTIMATED — adjust if wrong)
local RP_SIZE         = 0x58  -- sizeof(RecvProp) on x64 (estimated)

local DPT_NAMES = {
  [0]="int",[1]="float",[2]="vec3",[3]="vec2",
  [4]="char*",[5]="array",[6]="RecvTable*",[7]="int64",
}

-- ---- helpers ----
local function readQ(addr) return readQword(addr) or 0 end
local function readI(addr) return readInteger(addr) or 0 end
local function readStr(addr, n) return readString(addr, n or 64) or "" end

-- ---- get client.dll base ----
local clientBase = getAddress("client.dll")
if not clientBase or clientBase == 0 then
  print("ERROR: can't find client.dll — is CE attached to the game?")
  return
end
print(string.format("client.dll: 0x%X", clientBase))

-- ---- AOB scan for known class name strings ----
-- find "CWorld\0" or "CCSPlayer\0" in memory, then find what points AT it
local searchStrings = {
  { bytes = "43 57 6F 72 6C 64 00",             name = "CWorld" },     -- "CWorld\0"
  { bytes = "43 43 53 50 6C 61 79 65 72 00",    name = "CCSPlayer" },  -- "CCSPlayer\0"
  { bytes = "44 54 5F 42 61 73 65 45 6E 74 69 74 79 00", name = "DT_BaseEntity" },
}

local ccHead = nil

for _, search in ipairs(searchStrings) do
  print(string.format("scanning for '%s'...", search.name))
  local results = AOBScan(search.bytes, "+X-W-C")
  if results and results.Count > 0 then
    print(string.format("  found %d hit(s)", results.Count))
    -- for each string hit, scan for a pointer TO it (the m_pNetworkName field)
    for i = 0, math.min(results.Count - 1, 9) do
      local strAddr = tonumber(results[i])
      -- build AOB for the pointer value (little-endian 8 bytes)
      local p1 = strAddr % 256
      local p2 = math.floor(strAddr / 256) % 256
      local p3 = math.floor(strAddr / 65536) % 256
      local p4 = math.floor(strAddr / 16777216) % 256
      local p5 = math.floor(strAddr / 4294967296) % 256
      local p6 = math.floor(strAddr / 1099511627776) % 256
      local p7 = math.floor(strAddr / 281474976710656) % 256
      local p8 = math.floor(strAddr / 72057594037927936) % 256
      local ptrAOB = string.format("%02X %02X %02X %02X %02X %02X %02X %02X",
                                   p1, p2, p3, p4, p5, p6, p7, p8)

      local ptrResults = AOBScan(ptrAOB, "+X-W-C")
      if ptrResults and ptrResults.Count > 0 then
        print(string.format("  found %d pointer(s) to '%s' string", ptrResults.Count, search.name))
        -- check each pointer hit — if it's a m_pNetworkName field,
        -- the ClientClass starts 0x10 bytes before it
        for j = 0, math.min(ptrResults.Count - 1, 4) do
          local ptrAddr = tonumber(ptrResults[j])
          local ccAddr = ptrAddr - CC_NETWORKNAME

          -- validate: does m_pNext look like a pointer? does m_pRecvTable?
          local nextCC = readQ(ccAddr + CC_NEXT)
          local recvTable = readQ(ccAddr + CC_RECVTABLE)
          local nameCheck = readStr(readQ(ptrAddr), 32)

          if nextCC > 0x10000 and recvTable > 0x10000 and nameCheck == search.name then
            print(string.format("  ** VALID ClientClass at 0x%X (name='%s') **", ccAddr, nameCheck))
            ccHead = ccAddr
            break
          end
        end
        if ccHead then break end
      end
      -- cleanup scan results object
      if ptrResults then ptrResults.Destroy() end
    end
    results.Destroy()
    if ccHead then break end
  else
    print("  no hits")
  end
end

if not ccHead then
  print("ERROR: could not find ClientClass head.")
  print("Make sure the game is in an active map (not main menu).")
  return
end

-- ---- walk the linked list ----
print(string.format("\n=== walking ClientClass list from 0x%X ===", ccHead))

local classes = {}
local cc = ccHead
local visited = {}
local count = 0

while cc and cc > 0x10000 and not visited[cc] and count < 200 do
  visited[cc] = true
  count = count + 1

  local namePtr  = readQ(cc + CC_NETWORKNAME)
  local rtPtr    = readQ(cc + CC_RECVTABLE)
  local nextCC   = readQ(cc + CC_NEXT)
  local classID  = readI(cc + CC_CLASSID)

  if namePtr == 0 or rtPtr == 0 then break end

  local className = readStr(namePtr, 64)
  if className == "" then className = "(unknown)" end

  -- read RecvTable
  local propsPtr  = readQ(rtPtr + RT_PROPS)
  local nProps    = readI(rtPtr + RT_NPROPS)
  local tnPtr     = readQ(rtPtr + RT_NETTABLENAME)
  local tableName = readStr(tnPtr, 64)
  if tableName == "" then tableName = className end

  -- walk RecvProp array
  local props = {}
  if propsPtr > 0 and nProps > 0 and nProps < 500 then
    for i = 0, nProps - 1 do
      local propAddr = propsPtr + (i * RP_SIZE)
      local varNamePtr = readQ(propAddr + RP_VARNAME)
      local recvType   = readI(propAddr + RP_RECVTYPE)
      local offset     = readI(propAddr + RP_OFFSET)

      local varName = readStr(varNamePtr, 64)
      if varName and varName ~= "" and not varName:match("^%d") and varName ~= "baseclass" then
        local typeStr = DPT_NAMES[recvType] or string.format("type_%d", recvType)
        props[#props + 1] = { name = varName, type = typeStr, offset = offset }
      end
    end
  end

  classes[#classes + 1] = {
    className = className,
    tableName = tableName,
    classID   = classID,
    props     = props,
  }

  print(string.format("  [#%d] %s (id=%d, %d props) DT: %s",
        count, className, classID, #props, tableName))

  cc = nextCC
end

print(string.format("\n=== %d classes dumped ===", #classes))

-- ---- write output file ----
local desktop = os.getenv("USERPROFILE") .. "\\Desktop\\"
local outPath = desktop .. "netvar_dump.txt"
local f = io.open(outPath, "w")
if not f then
  print("ERROR: can't write to " .. outPath)
  return
end

f:write(string.format("// CS:S v93 x64 Netvar Dump (CE Lua)\n"))
f:write(string.format("// Dumped: %s\n\n", os.date("%Y-%m-%d %H:%M:%S")))
f:write(string.format("// %d classes\n\n", #classes))

for _, cls in ipairs(classes) do
  f:write(string.format("class %s {\n", cls.className))
  f:write(string.format("public:\n"))
  f:write(string.format("    // DT: %s, ClassID: %d\n", cls.tableName, cls.classID))
  for _, prop in ipairs(cls.props) do
    f:write(string.format("    VAROFF(%s, %s, 0x%04X);\n", prop.name, prop.type, prop.offset))
  end
  f:write("};\n\n")
end
f:close()

print(string.format("=== written to %s ===", outPath))

-- ---- print key classes summary ----
local important = { CCSPlayer=1, CBaseEntity=1, CBasePlayer=1, CBaseAnimating=1,
                    CBaseCombatWeapon=1, CCSPlayerResource=1 }

print("\n=== key offsets ===")
for _, cls in ipairs(classes) do
  for imp in pairs(important) do
    if cls.className:find(imp) then
      print(string.format("\n  %s (%s):", cls.className, cls.tableName))
      for _, prop in ipairs(cls.props) do
        print(string.format("    %-40s %-10s 0x%04X", prop.name, prop.type, prop.offset))
      end
    end
  end
end

print("\n=== done ===")

File: dumper.py
#!/usr/bin/env python3
"""
CS:S v93 x64 External Netvar Dumper
====================================
Talks to the CE MCP bridge (TCP relay on 127.0.0.1:9876) to walk the Source
engine's ClientClass linked list and dump every netvar offset for every class.
No injection — pure external memory reading.

PREREQUISITES:
1. CS:S running (cstrike_win64.exe) with bots spawned
2. Cheat Engine attached to the game
3. ce_mcp_bridge.lua executed in CE (pipe listening)
4. ce_tcp_relay.py running on 127.0.0.1:9876

USAGE:
    python dumper.py

OUTPUT:
    netvar_dump.txt — complete offset table, like the internal UC dumper but external.

STRUCTURE LAYOUTS (Source SDK x64, derived from cssourcex64 internal source):
The struct layouts below are the x64 versions. In Source x64:
  - pointers are 8 bytes
  - ClientClass has: m_pCreateFn, m_pCreateEventFn, m_pNetworkName (char*),
    m_pRecvTable (RecvTable*), m_pNext (ClientClass*), m_ClassID (int)
  - RecvTable has: m_pProps (RecvProp*), m_nProps (int), m_pDecoder,
    m_pNetTableName (char*), m_bInitialized (bool)
  - RecvProp has: m_pVarName (char*), m_RecvType (int), m_nFlags,
    StringBuffer, ... m_Offset (int) at a known position
"""

import socket
import struct
import json
import sys
import time

RELAY_HOST = "127.0.0.1"
RELAY_PORT = 9876

# Source SDK x64 struct offsets (derived from cssourcex64 usage patterns).
# these are the byte offsets of each field within the struct, in x64.
CLIENTCLASS_OFFSETS = {
    "m_pCreateFn":       0x00,   # function pointer (8 bytes)
    "m_pCreateEventFn":  0x08,   # function pointer (8 bytes)
    "m_pNetworkName":    0x10,   # const char* (8 bytes) — the class name string
    "m_pRecvTable":      0x18,   # RecvTable* (8 bytes)
    "m_pNext":           0x20,   # ClientClass* (8 bytes) — linked list
    "m_ClassID":         0x28,   # int (4 bytes)
}
CLIENTCLASS_SIZE = 0x30  # 48 bytes, padded

RECVTABLE_OFFSETS = {
    "m_pProps":          0x00,   # RecvProp* (8 bytes) — array of props
    "m_nProps":          0x08,   # int (4 bytes) — count
    # padding (4 bytes on x64 for alignment)
    "m_pDecoder":        0x10,   # pointer
    "m_pNetTableName":   0x18,   # const char* — the DT_ name
    "m_bInitialized":    0x20,   # bool
}
RECVTABLE_SIZE = 0x28

# RecvProp x64 layout — this is the tricky one. The internal code uses
# prop->GetOffset() which is a virtual, but the raw offset field is at a
# known position. From the SDK: m_Offset is stored after the type/flags.
# The layout (x64, packed with alignment):
#   0x00: m_pVarName (char*)
#   0x08: m_RecvType (int, SendPropType)
#   0x0C: m_nFlags (int)
#   0x10: StringBuffer (char[16])
#   0x20: ... proxy fn pointers ...
#   m_Offset: we need to find this empirically (scan for known offset values)
#
# For now, we use the SDK's known layout. m_Offset in RecvProp is typically
# at +0x4C on x64 (after proxy pointers). This may need adjustment.
RECVPROP_OFFSETS = {
    "m_pVarName":        0x00,   # const char*
    "m_RecvType":        0x08,   # int (SendPropType: DPT_Int=1, DPT_Float=2, ...)
    "m_Offset":          0x4C,   # int — THE FIELD OFFSET WE WANT
}
RECVPROP_SIZE = 0x58  # estimated; may vary

# SendPropType enum
DPT_TYPES = {
    0: "DPT_Int",
    1: "DPT_Float",
    2: "DPT_Vector",
    3: "DPT_VectorXY",
    4: "DPT_String",
    5: "DPT_Array",
    6: "DPT_DataTable",
    7: "DPT_Int64",
}

# ---- CE Bridge Client ----

class CEBridge:
    def __init__(self):
        self.sock = None
        self.req_id = 0

    def connect(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.settimeout(30)
        self.sock.connect((RELAY_HOST, RELAY_PORT))
        return True

    def close(self):
        if self.sock:
            self.sock.close()
            self.sock = None

    def call(self, method, params=None):
        """Send one JSON-RPC request, get one response. Single connection, sequential."""
        if not self.sock:
            raise RuntimeError("not connected")
        self.req_id += 1
        req = {"jsonrpc": "2.0", "method": method, "params": params or {}, "id": self.req_id}
        data = json.dumps(req).encode("utf-8")
        header = struct.pack("<I", len(data))
        self.sock.sendall(header + data)
        # read response header
        resp_header = b""
        while len(resp_header) < 4:
            chunk = self.sock.recv(4 - len(resp_header))
            if not chunk:
                raise ConnectionError("bridge closed connection")
            resp_header += chunk
        resp_len = struct.unpack("<I", resp_header)[0]
        if resp_len <= 0 or resp_len > 32 * 1024 * 1024:
            raise RuntimeError(f"bad response length: {resp_len}")
        # read response body
        resp_body = b""
        while len(resp_body) < resp_len:
            chunk = self.sock.recv(resp_len - len(resp_body))
            if not chunk:
                raise ConnectionError("bridge closed mid-response")
            resp_body += chunk
        return json.loads(resp_body.decode("utf-8")).get("result")

    def read_mem(self, addr, size):
        r = self.call("read_memory", {"address": addr, "size": size})
        if not r or "bytes" not in r:
            return None
        return bytes(r["bytes"])

    def read_qword(self, addr):
        b = self.read_mem(addr, 8)
        return struct.unpack_from("<Q", b)[0] if b else 0

    def read_dword(self, addr):
        b = self.read_mem(addr, 4)
        return struct.unpack_from("<I", b)[0] if b else 0

    def read_int(self, addr):
        b = self.read_mem(addr, 4)
        return struct.unpack_from("<i", b)[0] if b else 0

    def read_string(self, addr, max_len=64):
        r = self.call("read_string", {"address": addr, "max_length": max_len})
        if not r:
            return ""
        return r.get("value", "")

    def get_module_base(self, modname):
        r = self.call("enum_modules", {"limit": 300})
        if not r:
            return 0
        for m in r.get("modules", []):
            if m["name"].lower() == modname.lower():
                return int(m["address"], 16)
        return 0

    def aob_scan(self, pattern_hex, max_results=20):
        r = self.call("aob_scan", {"pattern": pattern_hex, "max_results": max_results})
        if not r:
            return []
        return [int(a, 16) if isinstance(a, str) else a for a in r.get("addresses", [])]


# ---- Dumper ----

def find_clientclass_head(bridge, client_base):
    """
    Find the head of the ClientClass linked list.

    Strategy: AOB-scan for a known network name string (e.g. "CWorld"), then
    find what points AT that string address (a ClientClass.m_pNetworkName field).
    The ClientClass struct is 0x10 bytes before the network name pointer.

    Fallback: scan for "CCSPlayer" or "DT_BaseEntity" strings.
    """
    # Try scanning for known class name strings
    for name in [b"CWorld\x00", b"CCSPlayer\x00", b"DT_BaseEntity\x00"]:
        hex_pattern = " ".join(f"{b:02X}" for b in name)
        print(f"  scanning for '{name.decode().strip(chr(0))}'...")
        hits = bridge.aob_scan(hex_pattern, max_results=20)
        if hits:
            print(f"  found {len(hits)} hit(s)")
            return hits, name.decode().strip(chr(0))
    return [], None


def dump_netvars(bridge, output_file="netvar_dump.txt"):
    print("=== CS:S x64 External Netvar Dumper ===")
    print()

    # 1. get module bases
    client_base = bridge.get_module_base("client.dll")
    if not client_base:
        print("ERROR: can't find client.dll — is the game running?")
        return
    print(f"client.dll base: 0x{client_base:X}")

    # 2. find ClientClass strings
    print()
    print("=== finding ClientClass linked list head ===")
    string_hits, found_name = find_clientclass_head(bridge, client_base)
    if not string_hits:
        print("ERROR: could not find any class name strings.")
        print("The AOB scan might need the game in an active state (map loaded, bots spawned).")
        return

    # 3. for each string hit, find what pointer references it
    # (a ClientClass.m_pNetworkName points at this string)
    print()
    print(f"=== searching for pointers to '{found_name}' string ===")
    print(f"  (this finds the ClientClass struct that owns this network name)")

    # The string address is in heap or data segment. We need to find a pointer
    # TO it. We scan the client.dll data segment for the address value.
    for hit_addr in string_hits[:5]:
        print(f"  string at 0x{hit_addr:X}, scanning for pointers to it...")
        # Convert the address to an AOB pattern (little-endian qword)
        addr_bytes = struct.pack("<Q", hit_addr)
        addr_hex = " ".join(f"{b:02X}" for b in addr_bytes)
        ptr_hits = bridge.aob_scan(addr_hex, max_results=10)
        if ptr_hits:
            print(f"  found {len(ptr_hits)} pointer(s) to this string")
            for ptr_addr in ptr_hits[:3]:
                # ptr_addr is the address of the m_pNetworkName field
                # ClientClass struct starts 0x10 bytes before m_pNetworkName
                cc_addr = ptr_addr - CLIENTCLASS_OFFSETS["m_pNetworkName"]
                print(f"  -> ClientClass at 0x{cc_addr:X}")

                # verify: read m_pNext and check it's also a valid pointer
                next_ptr = bridge.read_qword(cc_addr + CLIENTCLASS_OFFSETS["m_pNext"])
                recv_table_ptr = bridge.read_qword(cc_addr + CLIENTCLASS_OFFSETS["m_pRecvTable"])
                print(f"     m_pNext: 0x{next_ptr:X}")
                print(f"     m_pRecvTable: 0x{recv_table_ptr:X}")

                if next_ptr and recv_table_ptr:
                    print(f"  ** looks like a valid ClientClass! Walking the linked list... **")
                    walk_and_dump(bridge, cc_addr, output_file)
                    return

    print("ERROR: could not find a valid ClientClass struct.")
    print("The struct layout offsets might need adjustment for this build.")


def walk_and_dump(bridge, head_cc, output_file):
    """
    Walk the ClientClass linked list starting at head_cc.
    For each class, read its RecvTable and dump every RecvProp.
    """
    classes = []
    cc = head_cc
    visited = set()
    count = 0

    print()
    print(f"=== walking ClientClass linked list from 0x{head_cc:X} ===")

    while cc and cc not in visited and count < 200:
        visited.add(cc)
        count += 1

        # read ClientClass fields
        name_ptr = bridge.read_qword(cc + CLIENTCLASS_OFFSETS["m_pNetworkName"])
        recv_table_ptr = bridge.read_qword(cc + CLIENTCLASS_OFFSETS["m_pRecvTable"])
        next_cc = bridge.read_qword(cc + CLIENTCLASS_OFFSETS["m_pNext"])
        class_id = bridge.read_int(cc + CLIENTCLASS_OFFSETS["m_ClassID"])

        if not name_ptr or not recv_table_ptr:
            print(f"  [#{count}] 0x{cc:X}: invalid name/table ptr, stopping")
            break

        class_name = bridge.read_string(name_ptr, 64)
        if not class_name:
            class_name = "(unknown)"

        # read RecvTable
        props_ptr = bridge.read_qword(recv_table_ptr + RECVTABLE_OFFSETS["m_pProps"])
        n_props = bridge.read_int(recv_table_ptr + RECVTABLE_OFFSETS["m_nProps"])
        table_name_ptr = bridge.read_qword(recv_table_ptr + RECVTABLE_OFFSETS["m_pNetTableName"])
        table_name = bridge.read_string(table_name_ptr, 64) if table_name_ptr else class_name

        print(f"  [#{count}] {class_name} (id={class_id}, {n_props} props) — DT: {table_name}")

        # walk RecvProp array
        props = []
        if props_ptr and n_props > 0 and n_props < 500:
            for i in range(n_props):
                prop_addr = props_ptr + (i * RECVPROP_SIZE)
                var_name_ptr = bridge.read_qword(prop_addr + RECVPROP_OFFSETS["m_pVarName"])
                recv_type = bridge.read_int(prop_addr + RECVPROP_OFFSETS["m_RecvType"])
                offset = bridge.read_int(prop_addr + RECVPROP_OFFSETS["m_Offset"])

                var_name = bridge.read_string(var_name_ptr, 64) if var_name_ptr else ""
                if not var_name or var_name[0].isdigit():
                    continue
                if var_name == "baseclass":
                    continue

                type_str = DPT_TYPES.get(recv_type, f"type_{recv_type}")
                props.append((var_name, type_str, offset))

        classes.append({
            "class_name": class_name,
            "table_name": table_name,
            "class_id": class_id,
            "props": props,
        })

        cc = next_cc

    print()
    print(f"=== dumped {len(classes)} classes ===")

    # write output
    with open(output_file, "w") as f:
        f.write(f"// CS:S v93 x64 Netvar Dump (external, via CE MCP bridge)\n")
        f.write(f"// Dumped: {time.strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"// {len(classes)} classes\n\n")

        for cls in classes:
            f.write(f"class {cls['class_name']} {{\n")
            f.write(f"public:\n")
            f.write(f"    // DT: {cls['table_name']}, ClassID: {cls['class_id']}\n")
            for var_name, type_str, offset in cls["props"]:
                f.write(f"    VAROFF({var_name}, {type_str}, 0x{offset:X});\n")
            f.write(f"}};\n\n")

    print(f"=== written to {output_file} ===")

    # also print a summary of the important classes
    print()
    print("=== key offsets summary ===")
    important = ["CCSPlayer", "CBaseEntity", "CBasePlayer", "CBaseAnimating",
                 "CBaseCombatWeapon", "CCSPlayerResource"]
    for cls in classes:
        for imp in important:
            if imp in cls["class_name"]:
                print(f"\n  {cls['class_name']} ({cls['table_name']}):")
                for var_name, type_str, offset in cls["props"]:
                    print(f"    {var_name:40s} {type_str:10s} 0x{offset:04X}")
                break


if __name__ == "__main__":
    bridge = CEBridge()
    try:
        print("connecting to CE bridge on 127.0.0.1:9876...")
        bridge.connect()
        print("connected!")
        print()
        dump_netvars(bridge)
    except Exception as e:
        print(f"ERROR: {e}")
        import traceback
        traceback.print_exc()
    finally:
        bridge.close()
        print("\ndone.")
